wiki/python/ESTIMATOR_GUIDE.md

```diff
@@ -0,0 +1,833 @@
+   - In reality, this isn't necessary for this example. The decorator will look for an argument named `"X"` or default to the first, non `self`, argument.
+ - It's not necessary to convert to `CumlArray` and casting with `to_output` before returning. This function directly returned a `cp.ndarray` object. Any array-like object can be returned.
+ - Specifying `get_output_dtype=True` in the decorator argument instructs the decorator to also calculate the dtype in addition to the output type.
+
+## Do's And Don'ts
+
+### **Do:** Add Return Typing Information to Estimator Functions
+
+Adding the return type to estimator functions will allow the `Base` meta-class to automatically decorate functions based on their return type.
+
+**Do this:**
+```python
+def fit(self, X, y, convert_dtype=True) -> "KNeighborsRegressor":
+def predict(self, X, convert_dtype=True) -> CumlArray:
+def kneighbors_graph(self, X=None, n_neighbors=None, mode='connectivity') -> SparseCumlArray:
+def predict(self, start=0, end=None, level=None) -> typing.Union[CumlArray, float]:
+```
+
+**Not this:**
+```python
+def fit(self, X, y, convert_dtype=True):
+def predict(self, X, convert_dtype=True):
+def kneighbors_graph(self, X=None, n_neighbors=None, mode='connectivity'):
+def predict(self, start=0, end=None, level=None):
+```
+
+### **Do:** Return Array-Like Objects Directly
+
+There is no need to convert the array type before returning it. Simply return any array-like object and it will be automatically converted
+
+**Do this:**
+```python
+def predict(self) -> CumlArray:
+   cp_arr = cp.ones((10,))
+
+   return cp_arr
+```
+
+**Not this:**
+```python
+def predict(self, X, y) -> CumlArray:
+   cp_arr = cp.ones((10,))
+
+   # Don't be tempted to use `CumlArray(cp_arr)` here either
+   cuml_arr = input_to_cuml_array(cp_arr, order="K").array
+
+   return cuml_arr.to_output(self._get_output_type(X))
+```
+
+### **Don't:** Use `CumlArray.to_output()` directly
+
+Using `CumlArray.to_output()` is no longer necessary except in very rare circumstances. Converting array types is best handled with `input_to_cuml_array` or `cuml.using_output_type()` when retrieving `CumlArrayDescriptor` values.
+
+**Do this:**
+```python
+def _private_func(self) -> CumlArray:
+   return cp.ones((10,))
+
+def predict(self, X, y) -> CumlArray:
+
+   self.my_cupy_attribute_ = cp.zeros((10,))
+
+   with cuml.using_output_type("numpy"):
+      np_arr = self._private_func()
+
+      return self.my_cupy_attribute_ + np_arr
+```
+
+**Not this:**
+```python
+def _private_func(self) -> CumlArray:
+   return cp.ones((10,))
+
+def predict(self, X, y) -> CumlArray:
+
+   self.my_cupy_attribute_ = cp.zeros((10,))
+
+   np_arr = CumlArray(self._private_func()).to_output("numpy")
+
+   return CumlArray(self.my_cupy_attribute_).to_output("numpy") + np_arr
+```
+
+### **Don't:** Perform parameter validation in `__init__()`
```

Comment:
I would call this "parameter modification" or "transformation" -- validation (checking) should be fine.

Reply:
Done